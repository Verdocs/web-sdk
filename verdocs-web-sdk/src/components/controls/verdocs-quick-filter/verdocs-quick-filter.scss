@import '../../../theme.scss';

verdocs-quick-filter {
  position: relative;
  font-family: $primary-font;

  .control {
    outline: 0;
    height: 34px;
    display: flex;
    cursor: pointer;
    border-radius: 4px;
    align-items: center;
    flex-direction: row;
    box-sizing: border-box;
    background: $verdocs-grey-4;
    transition: all 100ms ease 0s;
    border: 1px solid $verdocs-bg-1;
  }

  .title {
    padding: 0 8px;
    font-size: 16px;
    max-width: 100%;
    overflow: hidden;
    color: $label-color;
    white-space: nowrap;
    box-sizing: border-box;
    text-overflow: ellipsis;

    .prefix {
      display: inline;
      font-weight: 700;
      margin: 0 7px 0 0;
      color: $secondary-color;
    }
  }

  .arrow {
    width: 32px;
    height: 32px;
    border: none;
    outline: none;
    cursor: pointer;
    align-items: center;
    display: inline-flex;
    justify-content: center;
    background: $verdocs-grey-4;
    transition: color 150ms ease 0s;

    svg {
      transition: 0.3s;
      fill: $verdocs-bg-1;
    }
  }

  .separator {
    width: 1px;
    display: block;
    margin-top: 8px;
    margin-bottom: 8px;
    align-self: stretch;
    box-sizing: border-box;
    background-color: $verdocs-bg-1;
  }

  .items {
    display: none;
    z-index: 10000;
    position: absolute;
    border-radius: 4px;
    background: $verdocs-grey-4;
    box-shadow: rgba(0, 0, 0, 0.1) 0 0 0 1px, rgba(0, 0, 0, 0.1) 0 4px 11px;
  }

  .items[data-show] {
    display: block;
  }

  .option {
    width: 100%;
    border: none;
    display: block;
    font-size: 16px;
    text-align: left;
    background: none;
    padding: 10px 20px;
    color: $verdocs-bg-5;
  }

  .option:hover {
    cursor: pointer;
    background: #4c56cb3f;
  }

  .option[disabled] {
    color: $light-disabled-color;
  }

  .option[disabled]:hover {
    cursor: inherit;
    background: transparent;
    color: $light-disabled-color;
  }

  &.open {
    .items {
      display: block;
    }

    .arrow svg {
      transform: scaleY(-1);
    }
  }
}
