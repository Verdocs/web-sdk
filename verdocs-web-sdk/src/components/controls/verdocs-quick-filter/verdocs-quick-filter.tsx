import {Host} from '@stencil/core';
import {createPopper, Instance} from '@popperjs/core';
import {Component, Prop, Element, State, h, Event, EventEmitter} from '@stencil/core';

const DropdownArrow = `<svg height="20" width="20" viewBox="0 0 20 20" aria-hidden="true" focusable="false" class="css-8mmkcg"><path d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"></path></svg>`;

export interface IFilterOption {
  value: string;
  label: string;
  disabled?: boolean;
}

/**
 * Display a drop-down menu of quick filter options.
 */
@Component({
  tag: 'verdocs-quick-filter',
  styleUrl: 'verdocs-quick-filter.scss',
})
export class VerdocsQuickFilter {
  @Element()
  el: HTMLElement;

  private dropdownButton?: HTMLElement;
  private dropdownMenu?: HTMLDivElement;
  private popper?: Instance;

  /**
   * The menu options to display.
   */
  @Prop() options: IFilterOption[] = [];

  @Prop({reflect: true})
  label: string = 'Filter';

  @Prop({reflect: true})
  value: string = '';

  @Prop({reflect: true})
  placeholder: string = 'Select...';

  /**
   * Event fired when a menu option is clicked.
   * Web Component events need to be "composed" to cross the Shadow DOM and be received by parent frameworks.
   */
  @Event({composed: true}) optionSelected: EventEmitter<IFilterOption>;

  @State() open: boolean;

  // We need to do this to reliably disconnect the click-away listener
  clickListenerSymbol = new AbortController();
  componentDidLoad() {
    this.popper = createPopper(this.dropdownButton, this.dropdownMenu, {placement: 'bottom-start', modifiers: [{name: 'offset', options: {offset: [-1, 14]}}]});

    window.addEventListener('click', this.clickListener.bind(this), {signal: this.clickListenerSymbol.signal});
  }

  disconnectedCallback() {
    this.clickListenerSymbol.abort();
  }

  clickListener(e: any) {
    const outside = !(e.target == this.el || this.el.contains(e.target));
    if (outside) {
      this.hide();
    }
  }

  handleSelectOption(e: any, option: IFilterOption) {
    e.stopPropagation();
    this.optionSelected.emit(option);
    this.value = option.value;
    this.hide();
  }

  // See https://popper.js.org/docs/v2/tutorial/
  // What we're doing here is clearing event listeners when they aren't needed, to increase performance in lists
  showDropdown() {
    this.open = true;
    this.dropdownMenu.setAttribute('data-show', '');
    this.dropdownMenu.removeAttribute('aria-hidden');

    this.popper
      ?.setOptions(options => ({
        ...options,
        modifiers: [...options.modifiers, {name: 'eventListeners', enabled: true}],
      }))
      .catch(() => {});

    this.popper?.update().catch(() => {});
  }

  handleHideDropdown(e: any) {
    if (e?.target?.localName === 'button' && e?.target?.className === 'arrow') {
      // This event is fired when a menu element is clicked because the button "loses focus" then too
      return;
    }

    this.hide();
  }

  toggleDropdown(e: any) {
    e.stopPropagation();

    if (this.open) {
      this.hide();
    } else {
      this.showDropdown();
    }
  }

  hide() {
    this.open = false;
    this.dropdownMenu.removeAttribute('data-show');
    this.dropdownMenu.setAttribute('aria-hidden', 'true');
    this.popper
      ?.setOptions(options => ({
        ...options,
        modifiers: [...options.modifiers, {name: 'eventListeners', enabled: false}],
      }))
      .catch(() => {});
  }

  render() {
    const selectedOption = this.options?.find(option => option.value === this.value);

    // This is bassd on our current font and size and the widest character "W" being about 15px wide.
    // It's a crude way to reserve a balanced amount of space so the top-level control doesn't jump
    // around when the selected option changes.
    const longestOption = this.options?.reduce((prev, curr) => (prev.label.length > curr.label.length ? prev : curr));
    const minWidth = `${(longestOption?.label || 'FILLER').length * 15 + 110}px`;

    return (
      <Host class={{open: !!this.open}} style={{minWidth, display: 'inline-block'}}>
        <div class="control" onClick={e => this.toggleDropdown(e)} onBlur={e => this.handleHideDropdown(e)}>
          <div class="title" ref={el => (this.dropdownButton = el as HTMLElement)}>
            <span class="prefix">{this.label}:</span>
            {selectedOption ? selectedOption.label : this.placeholder}
          </div>

          <div style={{flex: '1'}} />

          <div class="separator" />

          <div class="arrow" innerHTML={DropdownArrow} aria-label="Open Menu" />
        </div>

        <div class="items" aria-hidden={!this.open} ref={el => (this.dropdownMenu = el as HTMLDivElement)} style={{minWidth}}>
          {this.options?.map(option =>
            option.label ? (
              <button onClick={e => this.handleSelectOption(e, option)} class={`option ${option.value === this.value ? 'selected' : ''}`} disabled={option.disabled}>
                {option.label}
              </button>
            ) : (
              <div class="separator" />
            ),
          )}
        </div>
      </Host>
    );
  }
}
