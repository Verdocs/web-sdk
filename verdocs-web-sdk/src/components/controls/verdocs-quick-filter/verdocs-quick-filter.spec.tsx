import { newSpecPage } from '@stencil/core/testing';
import { VerdocsQuickFilter } from './verdocs-quick-filter';

describe('verdocs-quick-filter', () => {
  const options = [
    { label: 'Option 1', value: '1' },
    { label: 'Option 2', value: '2', disabled: true }
  ];

  it('renders with default props (with options)', async () => {
    const page = await newSpecPage({
      components: [VerdocsQuickFilter],
      html: `<verdocs-quick-filter options='${JSON.stringify(options)}'></verdocs-quick-filter>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with options and value', async () => {
    const page = await newSpecPage({
      components: [VerdocsQuickFilter],
      html: `<verdocs-quick-filter label="Filter" value="1" placeholder="All" options='${JSON.stringify(options)}'></verdocs-quick-filter>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('emits optionSelected event on option click', async () => {
    const page = await newSpecPage({
      components: [VerdocsQuickFilter],
      html: `<verdocs-quick-filter options='${JSON.stringify(options)}'></verdocs-quick-filter>`,
    });

    const spy = jest.fn();
    page.root.addEventListener('optionSelected', spy);

    // Simulate option selection
    if ((page.rootInstance as any).optionSelected) {
      (page.rootInstance as any).optionSelected.emit(options[0]);
      expect(spy).toHaveBeenCalled();
    }
  });
});
