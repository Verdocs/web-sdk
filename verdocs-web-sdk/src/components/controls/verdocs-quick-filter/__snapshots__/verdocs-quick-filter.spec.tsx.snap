// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-quick-filter renders with default props (with options) 1`] = `
<verdocs-quick-filter value="" style="min-width: 40px; display: inline-block;">
  <div class="control">
    <div class="title">
      <span class="prefix">
        Filter:
      </span>
      Select...
    </div>
    <div style="flex: 1;"></div>
    <div class="separator"></div>
    <div aria-label="Open Menu" class="arrow">
      <svg aria-hidden="true" class="css-8mmkcg" focusable="false" height="20" viewBox="0 0 20 20" width="20">
        <path d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"></path>
      </svg>
    </div>
  </div>
  <div aria-hidden="" class="items" style="min-width: 40px; position: absolute; left: 0; top: 0; margin: 0;">
    <button class="option">
      Option 1
    </button>
    <button class="option" disabled="">
      Option 2
    </button>
  </div>
</verdocs-quick-filter>
`;

exports[`verdocs-quick-filter renders with options and value 1`] = `
<verdocs-quick-filter value="1" style="min-width: 40px; display: inline-block;">
  <div class="control">
    <div class="title">
      <span class="prefix">
        Filter:
      </span>
      Option 1
    </div>
    <div style="flex: 1;"></div>
    <div class="separator"></div>
    <div aria-label="Open Menu" class="arrow">
      <svg aria-hidden="true" class="css-8mmkcg" focusable="false" height="20" viewBox="0 0 20 20" width="20">
        <path d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"></path>
      </svg>
    </div>
  </div>
  <div aria-hidden="" class="items" style="min-width: 40px; position: absolute; left: 0; top: 0; margin: 0;">
    <button class="option selected">
      Option 1
    </button>
    <button class="option" disabled="">
      Option 2
    </button>
  </div>
</verdocs-quick-filter>
`;
